bottlerocket-os/bottlerocket
# Add kernel-6.1 sources and use it for -dev variants

**Issue number:** https://github.com/bottlerocket-os/bottlerocket/issues/2855

**Description of changes:**

Add the sources for Linux kernel-6.1 and enable it for all the dev variants.

The following todo list does not block review of what is already there (config analysis) and will be added soon.
TODO:

- [x] Adjust SELinux policies for new classes `user_namespace` and `io_uring` (See Testing section below)
- [x] Update lookaside cache
- [x] potentially disable candidate options (See Testing section below)
- [x] Test build of an external kmod using kmod-kit

**Testing done:**

* Launch a `aws-dev` image and see it booting successfully with the correct kernel:
```
bash-5.1# uname -a
Linux ip-10-1-171-244.eu-central-1.compute.internal 6.1.19 bottlerocket-os/bottlerocket#1 SMP PREEMPT_DYNAMIC Mon May 15 16:07:55 UTC 2023 x86_64 GNU/Linux
```

* Check for missing SELinux classes (See todo above):
```
May 16 13:22:07 localhost kernel: SELinux:  Class io_uring not defined in policy.
May 16 13:22:07 localhost kernel: SELinux:  Class user_namespace not defined in policy.
May 16 13:22:07 localhost kernel: SELinux: the above unknown classes and permissions will be denied
May 16 13:22:07 localhost kernel: SELinux:  policy capability network_peer_controls=1
May 16 13:22:07 localhost kernel: SELinux:  policy capability open_perms=1
May 16 13:22:07 localhost kernel: SELinux:  policy capability extended_socket_class=1
May 16 13:22:07 localhost kernel: SELinux:  policy capability always_check_network=0
May 16 13:22:07 localhost kernel: SELinux:  policy capability cgroup_seclabel=1
May 16 13:22:07 localhost kernel: SELinux:  policy capability nnp_nosuid_transition=1
May 16 13:22:07 localhost kernel: SELinux:  policy capability genfs_seclabel_symlinks=1
May 16 13:22:07 localhost kernel: SELinux:  policy capability ioctl_skip_cloexec=0
May 16 13:22:07 localhost systemd[1]: Successfully loaded SELinux policy in 324.808ms.
```

* check for config changes against existing 5.15 kernels on all dev variants:
```
config-aarch64-aws-dev-diff:	 89 removed, 221 added,   4 changed
config-aarch64-metal-dev-diff:	 89 removed, 222 added,   4 changed
config-x86_64-aws-dev-diff:	 82 removed, 232 added,   6 changed
config-x86_64-metal-dev-diff:	 82 removed, 234 added,   6 changed
config-x86_64-vmware-dev-diff:	 98 removed, 230 added,   7 changed
```

As expected with a major kernel update there is lots of changes to the kernel config. I have tried to organize them into buckets for easier consumption and review. All changes that can be mapped to a specific commit and upstream series have been annotated with the corresponding data and a short summary has been added to all options.

The change buckets are (ordered by decreasing interest):

* **Candidates for disabling** (6/468 config changes). These are candidates for explicit disabling. I currently do not really see a good use case for damon in Bottlerocket, so saving the build cycles might be a good idea. Similarly we should probably disable the advanced driver interfaces for `ICE` as we have done for other NIC drivers. Details in [diff_disable.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_disable-csv)
* **Downstream config changes Amazon Linux did** (32/468 config changes). These are mainly due to dropping downstream backports of some drivers and a downstream port of lustreFS as well as some smaller config changes. Non of these should be problematic. Details are in [diff_al.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_al-csv)
* **Changes that are not needed right away but might be interesting** (9/468 config changes). This includes some options around FIPS related proc fs files to advertise certain fips properties, some net vendors that provide modern 10G+ NICs that might be relevant for metal and the audit driver to add DM_INTEGRITY logging to the audit subsystem. Details in [diff_interesting.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_interesting-csv).
* **Unexplainable changes** (5/468 config changes). These are config changes we are seeing between 5.15 and 6.1 for architectures where these options should not have been available or set to the new value all along through automatic selection or architecture dependent invisibility. I have no good explanation for these. Details in [diff_wat.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_wat-csv).
* **Architecture and toolchain specific changes** (72/468 config changes). These are config changes where either a differentiation is made based on the architecture or where the used build toolchain will have an impact on how certain features can be used or have to be configured. Nothing we can or would do about these. Details can be found in [diff_architecture.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_architecture-csv) and [diff_toolchain.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_toolchain-csv).
* **New drivers that are not needed** (85/468 config changes). These are drivers for various hardware that we do not need. A lot of these is for certain embedded applications. Nothing really interesting in this bucket. Details can be found in [diff_new_drivers_not_needed.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_new_drivers_not_needed-csv).
* **New functionality I am confident we do not need** (47/468 config changes). This is a mix of various new functionalities, from debug features to certain subsystems to new counters that can be exposed through sysfs. I think that non of these would be relevant for Bottlerocket use cases. Details can be found in [diff_new_not_needed.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_new_not_needed-csv).
* **General config cleanup** (197/468). This is a mix of various janitorial changes to the kernels config system. This includes removals due to deprecation of features, renaming of config options, changing of visibility through changes in dependencies, refactoring with splitting/merging certain options. Details can be found in [diff_config_cleanup.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_config_cleanup-csv).
* **The uninteresting remainder** (15/468). A bunch of options I could not really match to any of the other buckets, but deemed not really interesting. Details in [diff_uninteresting.csv](https://gist.github.com/foersleo/87a42ae412aed31591e695a1ad7b0fec#file-diff_uninteresting-csv).


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.
